File: PAT_Basic_Level_by_c/1079 延迟的回文数/1079 延迟的回文数.c
/*
 * 题目：1079 延迟的回文数
 * 实现原理：
 *   对一个非回文数进行变换：将其与自身的逆序相加，重复此过程直到得到回文数。
 *   如果10次迭代后仍未得到回文数，则输出"Not found in 10 iterations."
 *   算法步骤：
 *   1. is_palindrome函数：判断字符串是否为回文
 *   2. reverse函数：将字符串逆序
 *   3. add函数：实现两个大整数相加（以字符串形式）
 *   4. 主循环：最多迭代10次，每次输出变换过程
 */
#include <stdio.h>
#include <string.h>

int is_palindrome(char *s) {
    int len = strlen(s);
    for (int i = 0; i < len / 2; i++) {     // 只比较前半部分与对称位置
        if (s[i] != s[len - 1 - i]) {
            return 0;                       // 发现不对称即非回文
        }
    }
    return 1;
}

void reverse(char *s, char *r) {
    int len = strlen(s);
    for (int i = 0; i < len; i++) {
        r[i] = s[len - 1 - i];              // 逆序拷贝到r
    }
    r[len] = '\0';
}

void add(char *a, char *b, char *c) {
    int len = strlen(a);
    int carry = 0, idx = 0;
    for (int i = len - 1; i >= 0; i--) {    // 从低位到高位逐位相加
        int sum = (a[i] - '0') + (b[i] - '0') + carry;
        c[idx++] = sum % 10 + '0';
        carry = sum / 10;                   // 进位
    }
    if (carry) {                            // 最高位仍有进位则追加
        c[idx++] = carry + '0';
    }
    c[idx] = '\0';
    int i = 0, j = idx - 1;
    while (i < j) {                         // 结果是逆序存储的，翻转回正序
        char t = c[i];
        c[i] = c[j];
        c[j] = t;
        i++;
        j--;
    }
}

int main() {
    char s[2001], r[2001], c[2001];     // s:当前数; r:s的逆序; c:相加结果
    scanf("%s", s);
    if (is_palindrome(s)) {             // 输入本身已是回文数，直接输出
        printf("%s is a palindromic number.\n", s);
        return 0;
    }
    for (int i = 0; i < 10; i++) {      // 最多进行10次迭代
        reverse(s, r);
        printf("%s + %s = ", s, r);
        add(s, r, c);
        printf("%s\n", c);
        strcpy(s, c);                   // 用相加结果替换当前数
        if (is_palindrome(s)) {
            printf("%s is a palindromic number.\n", s);
            return 0;
        }
    }
    printf("Not found in 10 iterations.\n");    // 10次内未得到回文数
    return 0;
}

File: PAT_Basic_Level_by_java/1079 延迟的回文数/1079 延迟的回文数.java
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.io.IOException;

/**
 * 1079 延迟的回文数
 * 
 * 题目描述：
 * 给定一个正整数（不超过1000位），通过将其自身与其逆转数相加的方式，在10次迭代内寻找回文数。
 * 如果在10次迭代内得到回文数，输出每次的相加过程及最终回文数；否则报告"Not found in 10 iterations."。
 * 
 * 实现原理：
 * 解题思路：
 * - 由于数字可能长达1000位，远超 long 范围，需使用字符串处理大整数
 * - 每次迭代：将当前数字与其反转相加，判断结果是否为回文数
 * - 最多迭代10次，若找到回文数则输出并结束
 * 
 * 算法核心逻辑：
 * - isPalindrome()：双指针法判断字符串是否为回文（首尾字符逐一比较）
 * - reverse()：使用 StringBuilder.reverse() 实现字符串反转
 * - add()：模拟竖式加法，从最低位逐位相加，处理进位，最后反转结果
 * 
 * 时间复杂度：O(k×n)，k 为迭代次数（≤10），n 为数字长度。每次反转和加法均为 O(n)。
 * 空间复杂度：O(n)，用于存储相加结果的 StringBuilder。
 */
public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        String n = br.readLine().trim();
        
        // 判断初始输入是否已经是回文数
        if (isPalindrome(n)) {
            System.out.println(n + " is a palindromic number.");
            return;
        }
        
        // 最多迭代10次，每次将当前数与其反转相加
        for (int i = 0; i < 10; i++) {
            String a = n;
            String b = reverse(a);
            String c = add(a, b);
            System.out.println(a + " + " + b + " = " + c);
            // 判断结果是否为回文数
            if (isPalindrome(c)) {
                System.out.println(c + " is a palindromic number.");
                return;
            }
            n = c;
        }
        
        // 10次迭代仍未找到回文数
        System.out.println("Not found in 10 iterations.");
    }
    
    private static boolean isPalindrome(String s) {
        int left = 0, right = s.length() - 1;
        while (left < right) {
            if (s.charAt(left) != s.charAt(right)) {
                return false;
            }
            left++;
            right--;
        }
        return true;
    }
    
    private static String reverse(String s) {
        return new StringBuilder(s).reverse().toString();
    }
    
    // 大整数加法：从最低位逐位相加，处理进位
    private static String add(String a, String b) {
        StringBuilder sb = new StringBuilder();
        int carry = 0;  // 进位标志
        int i = a.length() - 1;
        int j = b.length() - 1;
        // 从末位开始逐位相加，直到所有位处理完毕且无进位
        while (i >= 0 || j >= 0 || carry > 0) {
            int d1 = i >= 0 ? (a.charAt(i) - '0') : 0;
            int d2 = j >= 0 ? (b.charAt(j) - '0') : 0;
            int sum = d1 + d2 + carry;
            sb.append(sum % 10);  // 当前位结果
            carry = sum / 10;     // 进位到高位
            i--;
            j--;
        }
        // 结果逆序，因为是从低位开始追加的
        return sb.reverse().toString();
    }
}
